react-components/src/components/RuleBasedOutputs/utils.ts

```diff
@@ -0,0 +1,218 @@
+      break;
+    }
+    case 'greaterThanOrEquals': {
+      const parameter = condition.parameters[0];
+      expressionResult = Number(asset[trigger.type]?.[trigger.key]) >= parameter;
+      break;
+    }
+    case 'within': {
+      const lower = condition.lowerBoundInclusive;
+      const upper = condition.upperBoundInclusive;
+      const value = Number(asset[trigger.type]?.[trigger.key]);
+      expressionResult = lower < value && value < upper;
+      break;
+    }
+    case 'outside': {
+      const lower = condition.lowerBoundExclusive;
+      const upper = condition.upperBoundExclusive;
+      const value = Number(asset[trigger.type]?.[trigger.key]);
+      expressionResult = value <= lower && upper <= value;
+      break;
+    }
+  }
+
+  return expressionResult;
+};
+
+const traverseExpression = (asset: Asset, expressions: Expression[]): boolean[] => {
+  let expressionResult: boolean = false;
+
+  const expressionResults: boolean[] = [];
+
+  expressions.forEach((expression) => {
+    switch (expression.type) {
+      case 'or': {
+        const operatorResult = traverseExpression(asset, expression.expressions);
+        expressionResult = operatorResult.find((result) => result) ?? false;
+        break;
+      }
+      case 'and': {
+        const operatorResult = traverseExpression(asset, expression.expressions);
+        expressionResult = operatorResult.find((result) => !result) ?? false;
+        break;
+      }
+      case 'not': {
+        const operatorResult = traverseExpression(asset, [expression.expression]);
+        expressionResult = !operatorResult[0];
+        break;
+      }
+      case 'numericExpression': {
+        expressionResult = checkNumericExpressionStatement(asset, expression);
+        break;
+      }
+      case 'stringExpression': {
+        expressionResult = checkStringExpressionStatement(asset, expression);
+        break;
+      }
+    }
+    expressionResults.push(expressionResult);
+  });
+
+  return expressionResults;
+};
+
+export const generateRuleBasedOutputs = (
+  model: CogniteCadModel,
+  contextualizedAssetNodes: Asset[],
+  assetMappings: AssetMapping3D[],
+  ruleSet: RuleOutputSet
+): void => {
+  const outputType = 'color'; // for now it only supports colors as the output
+
+  type RuleAndStyleIndex = {
+    styleIndex: TreeIndexNodeCollection;
+    ruleOutputParams: RuleOutput;
+  };
+
+  const ruleWithOutputs = ruleSet?.rulesWithOutputs;
+
+  ruleWithOutputs?.forEach((ruleWithOutput: { rule: Rule; outputs: RuleOutput[] }) => {
+    const { rule, outputs } = ruleWithOutput;
+    // Starting Expression
+    const expression = rule.expression;
+
+    const outputSelected = outputs.find(
+      (output: { type: string }) => output.type === outputType
+    ) as ColorRuleOutput;
+
+    if (outputSelected === undefined) return;
+
+    const ruleOutputAndStyleIndex: RuleAndStyleIndex = {
+      styleIndex: new TreeIndexNodeCollection(),
```

Comment:
I think a better way to construct this node collection, is to use the `AssetNodeCollection` that is provided by `Reveal`. You send it a bunch of asset IDs and it fetches the relevant data.

Reply:
Right. But I believe it would be safer to leave it the way it is because from what I quickly saw of the AssetNodeCollection, I may have to refactor more parts throughout the code and I'm afraid of ending up breaking more things right before the release.
I think it's safer to make this specific change more calmly over the next few days and after this release. Is it okay if we do like this?

Reply:
Yes, we can do this later :+1: